### PR TITLE
Biomarkers transform for ModelAD

### PR DESCRIPTION
**Pull request by beatrizsaldana**
Creates a new transform for the biomarkers dataset. The transform will re-structure data as described in [this](https://sagebionetworks.jira.com/browse/MG-99) jira ticket.

This is my first PR in this repo. Please review carefully and be as brutally honest as is necessary. It's better for me to learn things now than for us to have to go back and fix or add things later because nobody wanted to tell me I was doing something sub-optimally.

### Expected Changes
1. Added modelad_test_config.yaml
2. Added a biomarkers transform function 
3. Added test cases and test data

### Unexpected Changes
1. The `transform_biomarkers()` function outputs the transform as a `list` instead of `dict` or `pd.DataFrame` as is expected.
2. I added a `list_to_json()` function in `src/agoradatatools/etl/load.py` to acomodate the new output type
3. I added `elif isinstance(df, list):` in the `process_dataset()` function in `src/agoradatatools/process.py`.

@BWMac what do you think about the **Unexpected Changes**? Would it be better for the `transform_biomarkers()` function to output a `dict` or `pd.DataFrame` and prevent any of these extra changes? All feedback is welcome.

## UPDATE
No unexpected changes were implemented.

### CODE REVIEW

File: src/agoradatatools/process.py
Comment (beatrizsaldana) [inline src/agoradatatools/process.py]: @BWMac here are the changes mentioned in the PR description. What do you think about them?

File: src/agoradatatools/process.py
Comment (beatrizsaldana) [inline src/agoradatatools/process.py]: What do we think about this error handling? Is it necessary? Since we control all of the outputs, maybe we don't need it and can keep using the `else:` for the case where `df` is a `pd.DataFrame`.

Comment (BryanFauble) [reply src/agoradatatools/process.py]: I'll let Brad or others talk about the necessity of this, my comment is on the message itself.

If I saw this message "in a vacuum" I would ask myself "What WAS the Dataframe type when the exception was raised?"

could that information be added to the exception?

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (BryanFauble) [inline src/agoradatatools/etl/transform/biomarkers.py]: nit, are there more specific typing we can add for this aside from a generic `dict` or `list`?

Also - The docstring:
```
    Returns:
        dict: a dictionary of biomarkers data modeled after intended final JSON structure
```

Does not match the return type of the function of `list`. 


An example of how to add more specific typing (Might not be correct based on my comment above):

```
from typing import Dict, List

def transform_biomarkers(datasets: Dict[str, str]) -> List[str]:
```

File: src/agoradatatools/etl/load.py
Comment (BryanFauble) [inline src/agoradatatools/etl/load.py]: Generally, a context managed open is preferred like:

```
    with open(os.path.join(staging_path, filename), "w+") as temp_json:
          json.dump(df, temp_json, cls=NumpyEncoder, indent=2)
          return temp_json.name
```

This is so you don't need to be concerned about calling `.close()`, which is a valid way of accomplishing this, however, if this is the approach you want to take the `.close()` should be within a `finally` block so it's guaranteed to execute.

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (BryanFauble) [inline src/agoradatatools/etl/transform/biomarkers.py]: Should we add what the `biomarkers_dataset` is to the exception message?

File: src/agoradatatools/process.py
Comment (beatrizsaldana) [reply src/agoradatatools/process.py]: Really great point! I updated the error message, let me know if you think I should make any more changes here.

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/biomarkers.py]: Oops! Thank you for catching this! I updated the typing hints and the docstring for the function :)

File: src/agoradatatools/etl/load.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/load.py]: Hmm, I do like this approach better than what I was doing. I was trying to copy what the other functions are doing. Feedback please: Should I...
1. Update just this one function with the preferred context managed open
2. Update all of the X to json functions with the preferred context managed open
3. Leave things as they are and create a Jira ticket for updating the functions to use the preferred context managed open

Thoughts? @BryanFauble

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/biomarkers.py]: Yes we should! Thank you :)
I also changed the exception to TypeError because that is what it is.

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/biomarkers.py]: @BryanFauble @beatrizsaldana If I recall from previous experience, this isn't possible to do at this time. Adding type hints like `list[str]` isn't supported by Python 3.8 so our CI fails on it. Unfortunately we have to leave it generic.

Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/biomarkers.py]: Ohh, I have been using Python 3.9. I'll revert the typing hints back to generic dict, list, etc. Thank you! :)

Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/biomarkers.py]: You're fine to keep using 3.9, I think it's what gets installed by default when installing agora data tools, but our CI runs the test suite on 3(?) different Python versions, including 3.8, so sometimes we run across stuff that fails in 3.8 but not 3.9. I've run into this type hinting issue before haha.

Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/biomarkers.py]: Thank you! You probably saved me a lot of future failed debugging attempts :)

Comment (thomasyu888) [reply src/agoradatatools/etl/transform/biomarkers.py]: By the way, Python 3.8 EOL is coming up, so we may want to update

File: tests/transform/test_biomarkers.py
Comment (jaclynbeck-sage) [inline tests/transform/test_biomarkers.py]: I'm not sure I see a need to test both real data and fake data if they're both good input. Usually for my tests I just subset to a small number of rows from the real data as my test input, and then tweak a few things from there  if I need to check what happens with missing values or duplicates.

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (BryanFauble) [reply src/agoradatatools/etl/transform/biomarkers.py]: @jaclynbeck-sage @beatrizsaldana 
It is supported in Python 3.8 but not the way you are suggesting: `Adding type hints like list[str] isn't supported by Python 3.8 so our CI fails on it`

The following does work in Python 3.8:
```
from typing import Dict, List

def transform_biomarkers(datasets: Dict[str, str]) -> List[str]:
```

File: src/agoradatatools/etl/load.py
Comment (BryanFauble) [reply src/agoradatatools/etl/load.py]: I would:

1. Update any of the code you are already touching to following this approach
2. Log a tech debt ticket to go back and look at the other areas of the code

Generally, the mantra I follow is: "Leave the code in a better place than when I started". That needs to be balanced with the scope of the change, the time you have to make the changes, and the time it's going to take to validate the change. Some minor things are probably not worth fixing if it means there is a significant effort required to test the change.

File: tests/transform/test_biomarkers.py
Comment (jaclynbeck-sage) [inline tests/transform/test_biomarkers.py]: In your transform you have 2 error conditions: A TypeError and a ValueError. Can you make some test data that will force each of these to happen in a test? `test_proteomics_distribution_data.py` has an example of failure condition code that checks for different error types depending on input.

File: src/agoradatatools/etl/load.py
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/load.py]: I agree, update your own code and make a ticket for anything else you notice. I'm not sure who to assign the issue to so it doesn't get lost in the ether, maybe Jess?

File: tests/transform/test_biomarkers.py
Comment (jaclynbeck-sage) [reply tests/transform/test_biomarkers.py]: Actually now that I look at it, I'm not sure you _can_ trigger the `TypeError` condition, since the input _has_ to be a data frame by the time it gets to your transform. If it wasn't a DF, that means one of the earlier functions (load or standardize) would have failed first. In which case, it might be worth just removing that error check. `test_genes_biodomains.py` has a simpler, one-error-type failure case you can look at in that case.

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/biomarkers.py]: Ohh interesting. Learned something new! Thanks Bryan.

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/biomarkers.py]: See my comment on your test function, this error check probably isn't necessary.

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/biomarkers.py]: It might be worth changing this check from `==` to checking that biomarkers contains those columns, so that the data set has to have those columns in it but can have extra columns we don't care about.

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/transform/biomarkers.py]: We do have some code in place to basically do your transform in the form of a data frame instead of a list, where each row is a grouping and it contains a column with 'nested' data (it's the function `nest_fields` in `utils.py`). See the `genes_biodomains.py` transform for an example of how this is called. So in my head your grouping would be `['model', 'type', 'ageDeath', 'tissue', units']`, and you would want it to nest `['genotype', 'measurement', 'sex']` into a new column called `points`. However, we've only used/tested `nest_fields` with _one_ column as the grouping, never with multiple columns like is needed here, so I'm not sure this will work here. It might be simpler to leave the code as you've written it for now, take another look at `nest_fields` at some point to make it more generic to any number of columns, and edit later, but maybe @JessterB can weigh in.

File: src/agoradatatools/etl/load.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/load.py]: Thank you both for the feedback! I'll update the function I wrote, create a tech debt Jira ticket and assign it to Jess :)

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/biomarkers.py]: I'll use the typing library since I do really like to be specific with type hinting.

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (JessterB) [reply src/agoradatatools/etl/transform/biomarkers.py]: My 2c: I forgot about `nest_fields`... ideally we would use `nest_fields` for this, but if that will add a significant amount of work then we shouldn't tackle it as part of this PR. We can add a ticket to the backlog for extending `nest_fields` and move forward with the existing implementation.

File: src/agoradatatools/etl/load.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/load.py]: @JessterB I need help figuring out where to create this Jira ticket 🙃

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/biomarkers.py]: Yes, I was thinking about this earlier. The type hints should catch this :)

I'll remove it. Thank you for the validation!

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/biomarkers.py]: So true! I was trying to be strict with the error handling, but if there is a possibility for extra columns that we could just ignore, then I'll use `isin` or something like that instead of the `==`. Thank you!

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/biomarkers.py]: I did consider this approach but felt it was extra computational work that could be simplified with keeping the data as a list. However, I do agree that in terms of codebase maintenance it would be best for us to use existing approaches, because I did have to make a few changes to accommodate the new type of transform output. I can definitely update the transform to output a `pd.DataFrame` instead of a `list`. Thoughts?

**Comment by BWMac** [comment]
Just dropping a comment to say that I'm watching this PR and will do a review once it is marked as ready (unless told otherwise)!

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (BWMac) [inline src/agoradatatools/etl/transform/biomarkers.py]: NIT: This could be simplified to something like
```
missing_columns = [col for col in expected_columns if col not in biomarkers_dataset.columns]
if missing_columns:
    raise ValueError(f"Biomarker dataset missing columns: {', '.join(missing_columns)}")
```

File: src/agoradatatools/process.py
Comment (BWMac) [reply src/agoradatatools/process.py]: I think it would be preferable for the output of the new transform function to be one of the already supported data types. Do you think that would be feasible?

If it isn't feasible it's OK to add this. I will say that it was already a little confusing because we were referring to an object that could be a dictionary as `df` and adding lists into the mix adds to that confusion, so if this is the way we go maybe we can improve the variable naming while we are at it?

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (BWMac) [inline src/agoradatatools/etl/transform/biomarkers.py]: Just a thought (I don't feel strongly about this), but I don't think we typically include checks like this because this is essentially checking that the configuration file (which we are in control of) is correct. Do you think that this is necessary?

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (BWMac) [inline src/agoradatatools/etl/transform/biomarkers.py]: NIT: Is the `type_` variable a typo? and `ageDeath` should be in `camel_case`

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (BWMac) [inline src/agoradatatools/etl/transform/biomarkers.py]: It may be worthwhile to revisit this with the goal to 1. prevent using a nested loop if possible and 2. have less manually written out data structures (that we then have to maintain if something changes)

I was playing around with this a bit. Would something like this work?
```
data_as_list = []
group_columns = ["model", "type", "ageDeath", "tissue", "units"]
point_columns = ["genotype", "measurement", "sex"]
grouped = biomarkers_dataset.groupby(group_columns)

for group_key, group in grouped:
    entry = dict(zip(group_columns, group_key))
    entry["points"] = group[point_columns].to_dict('records')
    data_as_list.append(entry)
```

I tried it and the test passed, but I didn't scrutinize the data structure too closely.

File: src/agoradatatools/process.py
Comment (BWMac) [reply src/agoradatatools/process.py]: Yeah I don't think this is needed because of us being in control of the data structures that come out of the transform functions as you said @beatrizsaldana.

File: tests/transform/test_biomarkers.py
Comment (BWMac) [inline tests/transform/test_biomarkers.py:73]: The tests look good! But if you choose to keep both `ValueError` handling situations in the new transform, be sure to test the conditions needed to raise that error in both cases. Here you are only testing the second one I believe.

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/biomarkers.py]: I guess not haha! Thank you for the info, I'll remove it :)

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/biomarkers.py]: Its just a way to visually and programmatically avoid any conflicts with the built in `type` function.

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/transform/biomarkers.py]: This is a great idea, so much easier to maintain. Thank you!

File: src/agoradatatools/process.py
Comment (beatrizsaldana) [reply src/agoradatatools/process.py]: I also felt the variable `df` was a bit confusing, I'll change the variable name.

File: src/agoradatatools/process.py
Comment (beatrizsaldana) [inline src/agoradatatools/process.py]: @BWMac What do you think about this? I originally added a `raise Error` here but it turns out that the `neuropath_corr` dataset was causing our tests to fail because it was not a  `dict`, `list`, or `pd.DataFrame`. So I changed it to a warning. Thoughts?

File: src/agoradatatools/etl/transform/biomarkers.py
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/transform/biomarkers.py]: Agree, we don't need this check.

File: src/agoradatatools/process.py
Comment (beatrizsaldana) [reply src/agoradatatools/process.py]: Actually, I took a look and it seems that the variable `df` is being used a lot. Maybe it would be best I not touch it in this PR.

Feedback request: I kept the output data as a `list` because I felt it was most efficient. But if the priority is maintainability, I could transform it into a `pd.DataFrame` or a `dict`. Neither option is too much work and I would prefer to do what is best for the team, what do you think is best in this case (keep it as a list or change it to a df or dict)?

File: src/agoradatatools/etl/load.py
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/load.py]: I'll jump in: Go to JIRA (https://sagebionetworks.jira.com/), click on "Create" in the top bar, for project select "Model AD Explorer (MG)", Issue type = "Task" (Jess will change it if she wants something different), assign it to Jess, don't worry about filling in Team or Validator.

File: src/agoradatatools/process.py
Comment (jaclynbeck-sage) [reply src/agoradatatools/process.py]: What is the type of `neuropath_corr`? This seems odd.

Comment (jaclynbeck-sage) [reply src/agoradatatools/process.py]: Adding, `neuropath_corr` has never caused our tests to fail before, and calling `load.df_to_json` works on it, so now I'm extra puzzled.

File: src/agoradatatools/process.py
Comment (BWMac) [reply src/agoradatatools/process.py]: I think keeping it as a list is fine. You may want to add a `Union` type hint [here](https://github.com/Sage-Bionetworks/agora-data-tools/blob/d2b1ba03c5397250d2113e210c4947710c474cee/src/agoradatatools/process.py#L20) so that it is clear that the output of a transform can be any of these three data types.

File: src/agoradatatools/process.py
Comment (BWMac) [reply src/agoradatatools/process.py:144]: Yeah there is something weird going on here. Is this failing locally or in the CI/CD pipeline?

Comment (BWMac) [reply src/agoradatatools/process.py]: I see, the process tests are [failing](https://github.com/Sage-Bionetworks/agora-data-tools/actions/runs/11113747544/job/30878794890#step:5:805). This doesn't have anything to do with handling real data, just the [mock data](https://github.com/Sage-Bionetworks/agora-data-tools/blob/31c5806459ac5902f62cdf4dbd20e9c801a60bc6/tests/test_process.py#L69-L78) that I wrote into those unit tests.

I don't really think an error or warning here is that helpful, I would just add the list part and then have else be for dataframes as you said [here](https://github.com/Sage-Bionetworks/agora-data-tools/pull/148#discussion_r1776114408). I think the tests should pass that way.

Comment (BWMac) [reply src/agoradatatools/process.py:144]: I see. It was [failing](https://github.com/Sage-Bionetworks/agora-data-tools/actions/runs/11113747544/job/30878794890#step:5:805) in the GX tests in CI. When I run these locally, the warning you added isn't triggered...

File: src/agoradatatools/etl/load.py
Comment (JessterB) [reply src/agoradatatools/etl/load.py]: Thank you @jaclynbeck-sage, I missed this while I was out last week. What Jaclyn said works fine, once the ticket exists I can take it from there.

File: src/agoradatatools/process.py
Comment (beatrizsaldana) [reply src/agoradatatools/process.py]: It was failing because we use to have `if isinstance(df, dict): ...` `else: treat it as a pd.DataFrame`, which meant that the neuropath data was processed as a `pd.DataFrame`. The neuropath dataset it not of type `pd.DataFrame` but it did not matter because it was being treated as such. When I added the explicit `if isinstance(df, pd.DataFrame):` and the `else: raise Error`, it caused our tests to fail because the neuropath dataset would go straight into the `else: raise Error` block.

Comment (BWMac) [reply src/agoradatatools/process.py]: Right, but the pipeline was failing on the unit tests, not on the actual data processing. So it was failing because the strict type checking didn't like the mocked data structures that eventually made it through to that part. All I'm saying is that this was an issue that was specifically arising in the unit tests, not when handling real (or test) data.

File: src/agoradatatools/process.py
Comment (jaclynbeck-sage) [inline src/agoradatatools/process.py]: This comment isn't accurate... DataFrames will fall through to this else statement and we support them. I think it's fine to just remove the comment.

File: tests/transform/test_biomarkers.py
Comment (beatrizsaldana) [reply tests/transform/test_biomarkers.py]: Very true. I removed the "real" data test since it is hardest for us to visually validate.

File: src/agoradatatools/etl/load.py
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/load.py]: Nit: Can you add to the comment that the list needs to be a list of Dicts? Also, what happens if the list is a DataFrame? Does it still work (if you delete the typing specification)? Do we have a unit test for this function?

File: modelad_test_config.yaml
Comment (jaclynbeck-sage) [inline modelad_test_config.yaml]: Repeating Brad's comment from somewhere, ageDeath needs to be camel case: `age_death`. You'll have to change your test input files so this field is camel case too.

File: tests/transform/test_biomarkers.py
Comment (beatrizsaldana) [reply tests/transform/test_biomarkers.py:73]: Did not keep both `ValueError`s

File: src/agoradatatools/process.py
Comment (beatrizsaldana) [reply src/agoradatatools/process.py]: Exactly! Thank you for clarifying.

File: src/agoradatatools/process.py
Comment (beatrizsaldana) [reply src/agoradatatools/process.py]: Very true! Comment has been removed.

**Comment by BWMac** [comment]
@beatrizsaldana Check [this](https://github.com/Sage-Bionetworks/agora-data-tools/blob/dev/CONTRIBUTING.md#pre-commit-hooks) out if you haven't seen it yet for using `pre-commit` locally.

File: src/agoradatatools/etl/load.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/load.py]: We do not have a unit test for this function. I'll make one ASAP and include it in the PR.

File: src/agoradatatools/etl/load.py
Comment (jaclynbeck-sage) [inline src/agoradatatools/etl/load.py]: Looking at the `dict_to_json` function, it calls `remove_non_values` to clean up the dict before writing to JSON. I think this needs to be done to all the dicts in the list in this function. But... here's a question. Instead of outputting your transform as a `list`, what if you made it a `dict` where each item was another `dict` instead? This would bypass needing to add the `list_to_json` function + unit tests for it. I'm not sure if the JSON output will be identical but it might be worth testing?

File: src/agoradatatools/process.py
Comment (jaclynbeck-sage) [inline src/agoradatatools/process.py]: This needs to get removed per pre-commit.

File: src/agoradatatools/etl/load.py
Comment (jaclynbeck-sage) [reply src/agoradatatools/etl/load.py]: Before you do that, see my other comment in this file!

File: src/agoradatatools/etl/load.py
Comment (BWMac) [reply src/agoradatatools/etl/load.py]: This conversation has come up in a few places, just bringing my thoughts to this thread. 

Yesterday I [said](https://github.com/Sage-Bionetworks/agora-data-tools/pull/148#discussion_r1781720733) that it would be ideal if the data structure produced was one that we already support (dict or df) if feasible but I was [fine](https://github.com/Sage-Bionetworks/agora-data-tools/pull/148#discussion_r1781864436) with adding the ability to handle lists if that made the most sense for this new transform.

File: src/agoradatatools/process.py
Comment (BWMac) [inline src/agoradatatools/process.py]: NIT: the type hint should include `None` as a return option too

File: src/agoradatatools/etl/load.py
Comment (beatrizsaldana) [reply src/agoradatatools/etl/load.py]: Lets do it! I'll output it as a `dict` to make it easier for all of us to maintain the repo going forward (and aboid extra unit tests haha) :)

File: modelad_test_config.yaml
Comment (beatrizsaldana) [reply modelad_test_config.yaml]: I'm confused about this. Where do I need to make changes? The input csv column name is `age_death` and the Jira ticket outlining the output has it as `ageDeath`.

Comment (JessterB) [reply modelad_test_config.yaml]: you can change the name post-processing using `agora_rename` instead of `column_rename`, see e.g. the`neuropath_corr` dataset's config

Comment (JessterB) [reply modelad_test_config.yaml]: ...but actually, let's keep the field name as age_death in the json file. I think that means you can just omit both `column_rename` and `agora_rename` from the config. I'll explain why next time we sync, but I'll update your ticket now.

Comment (JessterB) [reply modelad_test_config.yaml]: this means you'll have to update this key in the expected json for the tests as well, sorry I didn't catch this sooner

**Review by thomasyu888 — APPROVED** [review]
🚀 LGTM!  I'll defer to others for final review, but the code looks great. Good looking tests!

File: modelad_test_config.yaml
Comment (BWMac) [reply modelad_test_config.yaml]: I will hold off on doing a final review until this change is implemented

**Review by JessterB — APPROVED** [review]
LGTM!

**Review by jaclynbeck-sage — APPROVED** [review]
Excellent! Everything looks great.

**Review by BWMac — APPROVED** [review]
🔥 LGTM

**Pull request merged**
